### PR TITLE
Update the myths in the Efficiency Guide for OTP 20

### PR DESCRIPTION
**Pull request by bjorng**
I originally started to do those updates for OTP 19, but didn't have time to finish them.

### CODE REVIEW

File: system/doc/efficiency_guide/myths.xml [hunk 1]
```diff
@@ -35,80 +35,29 @@
...
-    sure that the tail-recursive list function always is the fastest.</p>
+    using a tail-recursive function that builds a list in reverse
+    followed by a call to <c>lists:reverse/1</c> is faster than
+    a body-recursive function that builds the list in correct order.</p>
+
+    <p>That was true to some extent before R12B. It was even more true
+    before R7B. Today, not so much. It is generally not possible to
+    predict whether the tail-recursive or the body-recursive version will
+    be faster. Therefore, use the version that makes your code cleaner
+    (hint: it is usually the body-recursive version).</p>
```
Comment (essen) [inline system/doc/efficiency_guide/myths.xml]: Aren't body recursive functions more expensive in memory, and therefore not ideal when you are going to loop a large number of times? Might be worth mentioning if true.

File: system/doc/efficiency_guide/myths.xml [hunk 2]
```diff
@@ -199,6 +148,29 @@ vanilla_reverse([], Acc) ->
...
+
+    <p>Rewriting Erlang code to a NIF to make it faster should be
+    seen as a last resort. It is only guaranteed to be dangerous,
+    but not guaranteed to speed up the program.</p>
+
+    <p>Doing too much work in each NIF call will
+    <seealso marker="erts:erl_nif#WARNING">degrade responsiveness
+    of the VM</seealso>. Doing too little work may mean that
+    the gain of the faster processing in the NIF is eaten up by
+    the overhead of calling the BIF and checking the arguments.</p>
```
Comment (essen) [inline system/doc/efficiency_guide/myths.xml]: BIF? You lost me there.

Comment (bjorng) [inline system/doc/efficiency_guide/myths.xml]: Typo/thinko. Fixed.

File: system/doc/efficiency_guide/myths.xml (hunk 1 above)
Comment (bjorng) [inline system/doc/efficiency_guide/myths.xml]: > Aren't body recursive functions more expensive in memory, and therefore not ideal when you are going to loop a large number of times?

No, they are not more expensive in memory (any longer). I have pushed an update to mention that.

File: system/doc/efficiency_guide/retired_myths.xml [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  <section>
+    <title>Myth: List Comprehensions are Slow</title>
+
+    <p>List comprehensions used to be implemented using funs, and in the
+    old days funs were indeed slow.</p>
+
+    <p>Nowadays, the compiler rewrites list comprehensions into an ordinary
+    recursive function. Using a tail-recursive function with
+    a reverse at the end would be still faster. Or would it?
+    That leads us to the next myth.</p>
```
Comment (gomoripeti) [inline system/doc/efficiency_guide/retired_myths.xml]: there is no next myth in this chapter, or is there?

Comment (bjorng) [inline system/doc/efficiency_guide/retired_myths.xml]: Good catch! Fixed.

**Comment by vans163** [comment]
Just bumped into this 

```erlang
slower(<<Block:16/binary, Rest/binary>>, Acc) -> slower(Rest, Acc);
slower(_, Acc) -> Acc.
```
vs
```erlang
faster(<<>>, Acc) -> Acc;
faster(<<Block:16/binary, Rest/binary>>, Acc) -> faster(Rest, Acc).
```

Any word if this is a myth (expecting Rest rem 16 = 0).

**Comment by bjorng** [comment]
I am not sure I understand your example. Your functions returns `Acc`, but they never update `Acc`. Something missing in your example?

In other places in the Efficiency Guide it is said that more specific patterns can actually be faster than a variable that would match anything. Is that the point of your example?

**Comment by vans163** [comment]
Is it faster to have the empty pattern match first <<>> or is it faster to all through with _ once Rest = <<>>.

If the size of the binary is 0, 160, or 160,000,000.  I thought this was a common occurrence and logically one would think each case is checked in order, so if size of binary is 0, faster would be the fastest.

If size of binary is 160,000,000 then one would think slower would be the fastest as each recursion would skip the <<>> match.

Also I see what you are saying that Acc is never changed so this may be optimized away to just return Acc but I am sure you understood my idea, no?

**Comment by bjorng** [comment]
Yes, I understood your general idea, but the details are especially important for binary matching. The details can change how well the compiler can optimize binary matching.

I think that binary matching is too complicated to reduce to a single myth. There is an entire section in the Efficiency Guide about binaries: http://erlang.org/doc/efficiency_guide/binaryhandling.html

If I am concerned about performance of binary matching, I will always use the `bin_opt_info` or look at the generated BEAM assembly code.

**Comment by vans163** [comment]
@bjorng Thanks wow I was not aware of bin_opt_info, also that binaryhandling guide clears things up a lot.

**Pull request merged**